aips/aip-82.md

```diff
@@ -0,0 +1,149 @@
+
+## Impact
+
+The changes suggested herein make it easier for a unified approach for building applications that handle arbitrary types of fungible assets. Not approving of this places a burden on each project to instead create their own specialized dispatch for fungible assets that require this logic. This will substantially slow down adoption of assets which require stricter controls as each new asset would require an update to the dispatch table. Besides the maintenance cost, the dispatch table could also become incredibly inefficient over time due to loading too many external modules.
+
+## Alternative solutions
+
+* Require each provider to build their own specialized function and then have each provider build their own dispatch function. As mentioned above this quickly becomes untenable and not scalable.
+* Leverage dynamic dispatch of store creation. This is feasible, but we would rather receive feedback on this requirement rather than exposing more dynamic dispatch to the framework at this time. This would likely result in adhoc implementations that might not be secure.
+
+This solution unifies the solution and minimizes code.
+
+## Specification and Implementation Details
+
+This AIP introduces the following new functions:
+
+In `object.move`:
+```
+public fun root_owner<T: key>(object: Object<T>): address acquires ObjectCore {
+    let obj_owner = owner(object);
+    while (is_object(obj_owner)) {
+        obj_owner = owner(object::address_to_object<ObjectCore>(obj_owner));
+    };
+    obj_owner
+}
+```
+
+In `fungible_asset.move`:
+```
+/// Defnes a fungible asset stores as being untransferable at the object layer.
+#[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+struct Untransferable has key { }
+
+/// For untransferable stores allow deletion by the root owner.
+#[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+struct Deletable has key {
+    delete_ref: DeleteRef,
+}
+
+/// Set that only untransferable stores can be create for this fungible asset.
+public fun set_untransferable(constructor_ref: &ConstructorRef) {
+    let metadata_addr = object::address_from_constructor_ref(constructor_ref);
+    assert!(exists<Metadata>(metadata_addr), error::not_found(EFUNGIBLE_METADATA_EXISTENCE));
+    let metadata_signer = &object::generate_signer(constructor_ref);
+    move_to(metadata_signer, Untransferable {});
+}
+
+/// Creates an untransferable store.
+public fun create_untransferable_store<T>(account: &signer, metadata: Object<T>): Object<FungibleStore> {
```

Comment:
We should allow the creator to customize this object so the signature should be similar to `create_store`. So I believe the signature should be:
`public fun create_untransferable_store<T>(constructor_ref: ConstructorRef, metadata: Object<T>): Object<FungibleStore>`
Note now we move `ConstructorRef` into this function to ensure it cannot be used later to reset any flag.

Reply:
If we allow for `ConstructorRef`, we have no way of enforcing the non-transfer aspect.